catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/Poller.java

```diff
@@ -0,0 +1,270 @@
+            "Found {} entries to remove from the cache: {}",
+            keysToRemoveFromTheCache.size(),
+            StringUtils.join(keysToRemoveFromTheCache, ", "));
+      }
+      keysToRemoveFromTheCache
+          .stream()
+          .peek(key -> LOGGER.debug("Removing {} from the cache", key))
+          .forEach(cache::invalidate);
+    } else {
+      LOGGER.trace("Found no entries to remove from the cache");
+    }
+  }
+
+  /**
+   * Waits at most the given timeout for the loader process to complete
+   *
+   * @return a {@link Runnable} to commit the new value or handle the timeout or {@link
+   *     RuntimeException}
+   */
+  private Runnable getCommitRunnable(
+      final K key, final Future<V> loaderFuture, long timeout, TimeUnit timeoutTimeUnit) {
+    try {
+      final V newValue = loaderFuture.get(timeout, timeoutTimeUnit);
+      if (newValue == null) {
+        throw new IllegalStateException("The new value for " + key + " may not be null");
+      }
+
+      return () -> cacheNewValue(key, newValue);
+    } catch (TimeoutException e) {
+      loaderFuture.cancel(true);
+      return () -> handleTimeout(key);
+    } catch (ExecutionException e) {
+      final Throwable cause = e.getCause();
+      if (cause instanceof RuntimeException) {
+        return () -> handleException(key, (RuntimeException) cause);
+      } else {
+        throw new IllegalStateException("Exception when trying poll " + key, cause);
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new IllegalStateException("InterruptedException when trying to poll " + key, e);
+    }
+  }
+
+  /** @see #pollCurrent(long, TimeUnit, Map) */
+  abstract void handleTimeout(final K key);
+
+  /** @see #pollCurrent(long, TimeUnit, Map) */
+  abstract void handleException(final K key, final RuntimeException e);
+
+  /**
+   * Adds a entry to the cache or updates an existing value. Intended to be called by {@link
+   * #handleTimeout(Object)} and {@link #handleException(Object, RuntimeException)}.
+   */
+  /*final*/ synchronized void cacheNewValue(final K key, final V newValue) {
+    notNull(newValue);
+
+    final V nullableOldValue = cache.getIfPresent(key);
+
+    LOGGER.trace("Caching value={} for {}", newValue, key);
+    cache.put(key, newValue);
+
+    if (nullableOldValue != null) {
+      final V oldValue = nullableOldValue;
+      if (!java.util.Objects.equals(newValue, oldValue)) {
+        LOGGER.info("The value for {} was updated from {} to {}", key, oldValue, newValue);
+      }
+    } else {
+      LOGGER.debug("Polled {} for the first time. Found that the value is {}", key, newValue);
+    }
+  }
+
+  /** @throws IllegalArgumentException if the {@code executorService} is shutdown */
+  private static <T extends ExecutorService> T notShutdown(final T executorService) {
+    if (executorService.isShutdown()) {
+      throw new IllegalArgumentException("executorService may not be shutdown");
```

Comment:
❓ Did you mean `("executorService may be shutdown");`?

Reply:
It is unacceptable if the executorService is shutdown, so `executorService may not be shutdown` looks like the right phrase to me.